Catch parse errors and hand them to error_handler

anmonteiro opened the pull request:
refs #18.

This diff propagates parser errors to the request handling logic in order to give the user a chance to gracefully handle the error in `error_handler`. Previously httpaf was just swallowing these errors, e.g. `curl` would fail with `curl: (52) Empty reply from server` (this can be seen in the both echo examples in the repo).

I added tests for 2 cases I was seeing:
1. a malformed / rejected request, such as whitespace before a header's `:`, as per (comment in the codebase):
> From RFC7230§3.2.4:
>       "No whitespace is allowed between the header field-name and colon.
2. EOF in the HTTP Request, which mirrors the issue @sgrove reported in #18

seliopou commented on lib/server_connection.ml:
```diff
@@ -186,6 +186,7 @@ let set_error_and_handle ?request t error =
     t.error_handler ?request error (fun headers ->
       Writer.write_response writer (Response.create ~headers status);
       Body.of_faraday (Writer.faraday writer));
+    shutdown_writer t;
```
This will break streaming responses in that all write calls to the body will raise after `shutdown_writer` is called. It will also break asynchronous responses like in #18 since in that case the `Writer.write_response` call in callback passed to `error_handler` will be performed after `shutdown_writer` is called.

As separate test should be added for this case in a separate PR before this one moves forward. If the change itself is unnecessary, then it should also be removed.

anmonteiro replied:
Do you think replacing it with `wakeup_writer` (in case it works) is an option?

anmonteiro replied:
@seliopou It's been a little while, but IIRC one of the tests I added will break if this line is removed, so I'd say a different PR might be unnecessary.

seliopou replied:
I don't know if `wakeup_writer` will work here, which is why the existing, correct behavior should be tested independently before this PR is worked on further.

anmonteiro replied:
Ah, I understand. Thanks

seliopou replied:
Since the behavior that broke still isn’t tested, that behavior should be tested in a separate PR before this one is reviewed any further.

seliopou replied:
Suppose you changed nothing except for this line in the way you originally did. That would break existing behavior in the two ways I described. It may not be reachable, but it’s a big laying in wait.

So there should be a test for those cases independent of the other changes in this PR, and the it can be expanded to cover additional codepaths in this PR as necessary.

anmonteiro replied:
That's really helpful, thanks for dumbing it down for me. I'll do that in a separate PR

seliopou commented:
I haven't looked at the rest of the changes closely, but that bug immediately jumped out at me.

anmonteiro commented:
@seliopou Here's the current state of this PR:

- I've rebased on top of #102 
- I added 3 test cases that demonstrate the issues I'm seeing with http/af:
  1. A malformed request (space before colon in a header)
  2. A request EOF 
  3. Trying to stream a response from the error handler after handling a request EOF


The tentative fix in `Server_connection` fixes i. and ii., but not iii., which should fail in CI. Therefore it isn't the complete fix and I don't believe it's the correct one either. I do think, however, that the test cases successfully demonstrate what I'm seeing, so feel free to pull them into a PR of your own, or give me some guidance on how you'd like to see this fixed.
